hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -2917,3 +2918,115 @@ func aggregateMachineMessages(msgs []string) string {
 
 	return builder.String()
 }
+
+// secretJanitor reconciles secrets and determines which secrets should remain in the cluster and which should be cleaned up.
+// Any secret annotated with a nodePool name should only be on the cluster if the nodePool continues to exist
+// and if our current calculation for the inputs to the name matches what the secret is named.
+type secretJanitor struct {
+	*NodePoolReconciler
+}
+
+func (r *secretJanitor) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	secret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, req.NamespacedName, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error getting secret")
+		return ctrl.Result{}, err
+	}
+
+	nodePoolName, annotated := secret.Annotations[nodePoolAnnotation]
+	if !annotated {
+		return ctrl.Result{}, nil
+	}
+	log = log.WithValues("nodePool", nodePoolName)
+
+	nodePool := &hyperv1.NodePool{}
+	if err := r.Client.Get(ctx, supportutil.ParseNamespacedName(nodePoolName), nodePool); err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "error getting nodepool")
+		return ctrl.Result{}, err
+	} else if apierrors.IsNotFound(err) {
+		log.Info("removing secret as nodePool is missing")
+		return ctrl.Result{}, r.Client.Delete(ctx, secret)
+	}
+
+	hcluster, err := GetHostedClusterByName(ctx, r.Client, nodePool.GetNamespace(), nodePool.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	controlPlaneNamespace := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name)
+	expectedCoreConfigResources := expectedCoreConfigResourcesForHostedCluster(hcluster)
+	releaseImage, err := r.getReleaseImage(ctx, hcluster, nodePool.Status.Version, nodePool.Spec.Release.Image)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	config, missingConfigs, err := r.getConfig(ctx, nodePool, expectedCoreConfigResources, controlPlaneNamespace, releaseImage, hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if missingConfigs {
+		return ctrl.Result{}, nil
+	}
+	targetVersion := releaseImage.Version()
+
+	pullSecretName, err := r.getPullSecretName(ctx, hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	globalConfig, err := globalConfigString(hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	targetPayloadConfigHash := supportutil.HashSimple(config + targetVersion + pullSecretName + globalConfig)
```

Comment:
ideally this would be forced by a function signature or similar since we are using it on creation as well, otherwise there's a divergence risk

Reply:
Yeah, was hoping there would be a suggestion on this. They're all strings so even a signature ends up being stringly typed - WDYT?

Reply:
yeh not ideal but I guess having a signature is still slightly better, say you are modifying the code in the main loop, you add a new parameter. Binary won't compile unless you include the same number of parameters here. Then yeh you could satisfy it by introducing the wrong string here.
Maybe start with a signature and then refactor a bit towards having a single `func targetHash(nodePool) (string)`?

Reply:
`func targetHash(nodePool) string` is not really possible since getting all the data for it requires a lot of other actions - and, even if we had it hang off of the reconciler as a method, the main reconciliation method ends up using a lot of the intermediate calculated steps in other parts of the method, so it would be something like `func (r *NodePoolReconciler) targetHash(nodePool) (expectedCoreConfigResources, releaseImage, ... hash)`. I can do that but it seemed awkward.